cloud/scope/vnic_reconciler.go

```diff
@@ -0,0 +1,142 @@
+package scope
+
+import (
+	"context"
+	"fmt"
+
+	infrastructurev1beta1 "github.com/oracle/cluster-api-provider-oci/api/v1beta1"
+	"github.com/oracle/cluster-api-provider-oci/cloud/ociutil"
+	"github.com/oracle/oci-go-sdk/v65/common"
+	"github.com/oracle/oci-go-sdk/v65/core"
+	"github.com/pkg/errors"
+)
+
+func (m *MachineScope) ReconcileVnicAttachments(ctx context.Context) error {
+	if m.IsControlPlane() {
+		return errors.New("cannot attach multiple vnics to ControlPlane machines")
+	}
+
+	for index, vnicAttachment := range m.OCIMachine.Spec.VnicAttachments {
+		if m.vnicAttachmentExists(ctx, vnicAttachment) {
+			m.Logger.Info("vnicAttachment", vnicAttachment.DisplayName, " already exists")
+			m.updateVnicAttachment(ctx, vnicAttachment)
+			continue
+		}
+
+		vnicId, err := m.createVnicAttachment(ctx, vnicAttachment)
+		if err != nil {
+			msg := fmt.Sprintf("Error creating VnicAttachment %s for cluster %s",
+				*vnicAttachment.DisplayName, m.Cluster.Name)
+			m.Logger.Error(err, msg)
+			return err
+		}
+
+		m.OCIMachine.Spec.VnicAttachments[index].VnicAttachmentId = vnicId
+	}
+
+	return nil
+}
+
+func (m *MachineScope) createVnicAttachment(ctx context.Context, spec infrastructurev1beta1.VnicAttachment) (*string, error) {
+	vnicName := spec.DisplayName
+
+	// Default to machine subnet if spec doesn't supply one
+	subnetId := m.getWorkerMachineSubnet()
+	if spec.SubnetName != "" {
+		var err error
+		subnetId, err = m.getMachineSubnet(spec.SubnetName)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	tags := m.getFreeFormTags(*m.OCICluster)
+
+	definedTags := make(map[string]map[string]interface{})
+	if m.OCIMachine.Spec.DefinedTags != nil {
```

Comment:
can we use a common function for this at machine.go? or do we have one?

Reply:
fixed